### PR TITLE
6283 add a coop qbo migration

### PR DESCRIPTION
**Pull request by X0nic**

### CODE REVIEW

File: lib/tasks/migrate_data.rake [hunk 1]
```diff
@@ -99,4 +99,36 @@ namespace :tww do
       end
     end
   end
+
+  desc "migrate some test data from quickbooks"
```
Comment (hooverlunch) [inline lib/tasks/migrate_data.rake]: Please put this in a separate file unless there is some reason to keep it in this one.

File: lib/tasks/migrate_data.rake [hunk 2]
```diff
@@ -99,4 +99,36 @@ namespace :tww do
...
+  task migrate_organizations_to_qbo: :environment do
+    qb_connection = Division.root.qb_connection
+
+    customers = Quickbooks::Service::Customer.new(qb_connection.auth_details).all
+
+    customer_hash = {}
+    customers.each do |customer|
+      # Customer does not have a name property so we use display name since that is
+      # what is usually show in the QBO UI.
```
Comment (hooverlunch) [inline lib/tasks/migrate_data.rake]: Nice comment.

File: lib/tasks/migrate_data.rake [hunk 3]
```diff
@@ -99,4 +99,36 @@ namespace :tww do
...
+        if org.qb_id.blank?
+          puts "Mapping organization '#{org.name}' to #{customer.id}"
+          org.update!(qb_id: customer.id)
+        else
+          if org.qb_id != customer.id
+            puts "ERROR! 'Problem mapping customer #{customer.id} to Coop #{org.name}', it is already mapped to #{org.qb_id}, skipping"
+            next
+          end
+        end
+      end
```
Comment (hooverlunch) [inline lib/tasks/migrate_data.rake]: Should we have an `else` block in which we output a message that no matching Org was found for customer xxx?

File: lib/tasks/migrate_data.rake [hunk 4]
```diff
@@ -99,4 +99,36 @@ namespace :tww do
...
+    qb_connection = Division.root.qb_connection
+
+    customers = Quickbooks::Service::Customer.new(qb_connection.auth_details).all
+
+    customer_hash = {}
+    customers.each do |customer|
+      # Customer does not have a name property so we use display name since that is
+      # what is usually show in the QBO UI.
+      customer_hash[customer.display_name] = customer
+    end
```
Comment (hooverlunch) [inline lib/tasks/migrate_data.rake]: Note that the above could be written simply as `customers_hash = customers.index_by(&:display_name)`.

File: lib/tasks/migrate_data.rake [hunk 5]
```diff
@@ -99,4 +99,36 @@ namespace :tww do
...
+
+    Organization.all.find_each do |org|
+      customer = customer_hash[org.name]
+
+      if customer
+        if org.qb_id.blank?
+          puts "Mapping organization '#{org.name}' to #{customer.id}"
+          org.update!(qb_id: customer.id)
+        else
+          if org.qb_id != customer.id
```
Comment (hooverlunch) [inline lib/tasks/migrate_data.rake]: You can just use `elsif` here.

File: lib/tasks/migrate_data.rake [hunk 6]
```diff
@@ -99,4 +99,36 @@ namespace :tww do
...
+            puts "ERROR! 'Problem mapping customer #{customer.id} to Coop #{org.name}', it is already mapped to #{org.qb_id}, skipping"
```
Comment (hooverlunch) [inline lib/tasks/migrate_data.rake]: Nice error msg!

**Comment by emitche** [comment]
Per @X0nic in Redmine on 2017-06-06:
> One customer has a colon in the name. Qbo won't allow this.

Before merging, we need to address this. @linesarefuzzy to pair.

**Comment by emitche** [comment]
Update: Colon in customer name does not affect this issue. This issue can move forward.

**Pull request merged**